### PR TITLE
Added dequeue statistics

### PR DESCRIPTION
**Pull request by dominik-cnx**
With this PR we are able to monitor details regarding the dequeueing, namely

       lastDequeueAttemptTimestamp
       lastDequeueSuccessTimestamp
       nextDequeueDueTimestamp

This allows us to easily identify e.g. 
- an  overload situation where queues are not longer processed (dequeue is no longer attempted)
- dequeue was attempted but failed

### CODE REVIEW

**Review by mcweba — APPROVED** [review]
Nice addon :-)

**Review by Kusig — COMMENTED** [review]
Is there an explizit reason why this was not implemented as part of the existing QueueStatistics handler.
Note that the QueueStatistics are already consumed frequently by our monitoring system and therefore if extended there it would have been easier integrated within Grafana dashboards etc. as it exists already.

File: src/main/java/org/swisspush/redisques/RedisQues.java [hunk 1]
```diff
@@ -93,6 +94,14 @@ private enum QueueState {
 
     private Map<QueueOperation, QueueAction> queueActions = new HashMap<>();
 
+    private Map<String, DequeueStatistic> dequeueStatistic = new ConcurrentHashMap<>();
+
+    public class DequeueStatistic{
```
Comment (Kusig) [inline src/main/java/org/swisspush/redisques/RedisQues.java:99]: Why not making this part of the existing QueueStatistics instead of another one here?

Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/RedisQues.java:99]: The idea was to provide this to the UI of our  communication server that lists the queues. Let me see if we can also easily provide these values through the QueueStatisticsHandler / QueueStatisticsCollector

**Comment by mcweba** [comment]
PR is OK for me (when the conflicts have been resolved)

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java [hunk 2]
```diff
@@ -408,67 +446,67 @@ JsonObject getAsJsonObject() {
...
-        ctx.event = event;
         ctx.queueNames = queues;
-        step1(ctx);
+        step1(ctx, promise);
+        return promise.future();
     }
 
     /** <p>init redis connection.</p> */
-    void step1(RequestCtx ctx) {
+    void step1(RequestCtx ctx, Promise<JsonObject> promise) {
```
Comment (dominik-cnx) [inline src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:468]: Propose to return a Promise instead of passing one in

Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java]: And then somehow run them using the CompositeFuture

File: src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java [hunk 3]
```diff
@@ -452,34 +452,43 @@ JsonObject getAsJsonObject() {
...
+        final Promise<JsonObject> promise = Promise.promise();
         if (queues == null || queues.isEmpty()) {
             log.debug("Queue statistics evaluation with empty queues, returning empty result");
             promise.complete(new JsonObject().put(STATUS, OK).put(RedisquesAPI.QUEUES, new JsonArray()));
             return promise.future();
         }
         var ctx = new RequestCtx();
         ctx.queueNames = queues;
-        step1(ctx, promise);
+        step1(ctx).compose(
```
Comment (mcweba) [inline src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:463]: Why do you call all the steps here and not as before inside the method? So step1 could return step2(ctx)

Comment (dominik-cnx) [reply src/main/java/org/swisspush/redisques/util/QueueStatisticsCollector.java:463]: The idea is to have the steps independent of each other and chain them here at a centralized location.

**Pull request merged**